cumulusci/tasks/metadata/modify.py

```diff
@@ -1,52 +1,111 @@
+        self.elements = self.options["elements"]
+
     def _run_task(self):
-        chdir = self.options.get("chdir")
-        if chdir:
-            self.logger.info("Changing directory to {}".format(chdir))
-        with cd(chdir):
-            for element in self.options["elements"]:
+        if self.chdir:
+            self.logger.info("Changing directory to {}".format(self.chdir))
+        with cd(self.chdir):
+            for element in self.elements:
                 self._process_element(element)
 
-    def _process_element(self, element):
+    def _process_element(self, step):
         self.logger.info(
-            "Removing elements matching {xpath} from {path}".format(**element)
+            "Removing elements matching {xpath} from {path}".format(**step)
         )
-        for f in glob.glob(element["path"]):
+        for f in glob.glob(step["path"], recursive=True):
+            self.logger.info(f"Checking {f}")
             with open(f, "rb") as fp:
                 orig = fp.read()
             root = ET.parse(f)
-            res = root.findall(
-                element["xpath"].replace(
-                    "ns:", "{http://soap.sforce.com/2006/04/metadata}"
-                )
+            res = root.xpath(
+                step["xpath"],
+                namespaces={
+                    "ns": "http://soap.sforce.com/2006/04/metadata",
+                    "re": "http://exslt.org/regular-expressions",
+                },
             )
+            self.logger.info(f"Found {len(res)} matching elements")
             for element in res:
                 element.getparent().remove(element)
-            processed = (
-                b'<?xml version="1.0" encoding="UTF-8"?>\n'
-                + ET.tostring(root, encoding="utf-8")
-                + b"\n"
-            )
+
+            processed = bytes(salesforce_encoding(root), "utf-8")
+
             if orig != processed:
                 self.logger.info("Modified {}".format(f))
                 with open(f, "wb") as fp:
                     fp.write(processed)
+
+
+def has_content(element):
+    return element.text or list(element)
+
+
+def salesforce_encoding(xdoc):
+    r = xml_encoding
+    if SF_NS in xdoc.getroot().tag:
+        xdoc.getroot().attrib["xmlns"] = SF_NS
+    for action, elem in ET.iterwalk(
+        xdoc, events=("start", "end", "start-ns", "end-ns", "comment")
+    ):
+        if action == "start-ns":
+            pass  # handle this nicely if SF starts using multiple namespaces
+        elif action == "start":
+            tag = elem.tag
+            if "}" in tag:
+                tag = tag.split("}")[1]
+            text = (
+                escape(elem.text, {"'": "&apos;", '"': "&quot;"})
+                if elem.text is not None
+                else ""
+            )
+
+            attrs = "".join([f' {k}="{v}"' for k, v in elem.attrib.items()])
+            if not has_content(elem):
+                r += f"<{tag}{attrs}/>"
```

Comment:
This line is missing coverage

Reply:
fd28e00d92

Reply:
https://coveralls.io/builds/27582317/source?filename=cumulusci/tasks/metadata/modify.py#L107